### PR TITLE
[DPE-3881] Use ruff as a linter and formatter

### PR DESCRIPTION
**Pull request by shayancanonical**
## Issue
We should use ruff as a linter and formatter

## Solution
Use ruff. Fix lint and format warnings

### CODE REVIEW

**Review by paulomach — APPROVED** [review]
lgtm. Worth changing the `ConnectionError` class though

File: pyproject.toml [hunk 1]
```diff
@@ -85,20 +77,16 @@ log_cli_level = "INFO"
...
-max-doc-length = 99
-max-complexity = 10
-exclude = [".git", "__pycache__", ".tox", "build", "dist", "*.egg_info", "venv"]
-select = ["E", "W", "F", "C", "N", "R", "D", "H"]
+[tool.ruff.lint]
+explicit-preview-rules = true
+select = ["A", "E", "W", "F", "C", "N", "D", "I", "CPY001"]
 # Ignore W503, E501 because using black creates errors with this
 # Ignore D107 Missing docstring in __init__
 # Ignore D105 Missing docstring in magic method
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply pyproject.toml]: addressed in 871e5d986927f324bff23279d8f98b5d217d01e7

File: src/server_exceptions.py [hunk 2]
```diff
@@ -12,7 +12,7 @@ class Error(status_exception.StatusException):
     """MySQL Server unreachable or unhealthy"""
 
 
-class ConnectionError(Error):
+class ConnectionError(Error):  # noqa: A001
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply src/server_exceptions.py]: addressed in 871e5d986927f324bff23279d8f98b5d217d01e7

File: src/container.py [hunk 3]
```diff
@@ -177,7 +177,7 @@ def _run_command(
         command: typing.List[str],
         *,
         timeout: typing.Optional[int],
-        input: str = None,
+        input: str = None,  # noqa: A002
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply src/container.py]: addressed in 871e5d986927f324bff23279d8f98b5d217d01e7

File: src/snap.py [hunk 4]
```diff
@@ -261,7 +257,7 @@ def _run_command(
         command: typing.List[str],
         *,
         timeout: typing.Optional[int],
-        input: str = None,
+        input: str = None,  # noqa: A002
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply src/snap.py]: addressed in 871e5d986927f324bff23279d8f98b5d217d01e7

File: pyproject.toml [hunk 5]
```diff
@@ -85,34 +77,50 @@ log_cli_level = "INFO"
...
+explicit-preview-rules = true
+select = ["A", "E", "W", "F", "C", "N", "D", "I", "CPY001"]
+ignore = [
+    "D102", # Ignore D102 Missing docstring in public method
+    "D105", # Ignore D105 Missing docstring in magic method
+    "D107", # Ignore D107 Missing docstring in __init__
+    "D203", # Ignore D203 1 blank line required before class docstring
+    "D204", # Ignore D204 1 blank line required after class docstring
+    "D213", # Ignore D213 Multi-line docstring summary should start at the second line
+    "D215", # Ignore D215 Section underline is over-indented
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply pyproject.toml]: removed in 484fedacf91212f8c0c9003245ce79b0e1d21841

File: pyproject.toml [hunk 6]
```diff
@@ -85,34 +77,50 @@ log_cli_level = "INFO"
...
+    "D400", # Ignore D400 First line should end with a period
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply pyproject.toml]: removed in 484fedacf91212f8c0c9003245ce79b0e1d21841

File: pyproject.toml [hunk 7]
```diff
@@ -85,34 +77,50 @@ log_cli_level = "INFO"
...
+    "D403", # Ignore D403 First word of the first line should be capitalized
+    "D404", # Ignore D404 First word of the docstring should not be "This"
+    "D406", # Ignore D406 Section name should end with a newline
+    "D407", # Ignore D407 Missing dashed underline after section
+    "D408", # Ignore D408 Section underline should be in the line following section's name
+    "D409", # Ignore D409 Section underline should match the length of its name
+    "D413", # Ignore D413 Missing blank line after last section
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply pyproject.toml]: removed in 484fedacf91212f8c0c9003245ce79b0e1d21841

File: pyproject.toml [hunk 8]
```diff
@@ -85,34 +77,50 @@ log_cli_level = "INFO"
...
+    "D415", # Ignore D415 First line should end with a period, question mark, or exclamation point
+    "E501", # Ignore E501 Line too long
+    "N801", # Ignore N801 Class name should use CapWords convention
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply pyproject.toml]: added noqa in 484fedacf91212f8c0c9003245ce79b0e1d21841

File: pyproject.toml [hunk 9]
```diff
@@ -85,34 +77,50 @@ log_cli_level = "INFO"
...
+ignore = [
+    "D102", # Ignore D102 Missing docstring in public method
+    "D105", # Ignore D105 Missing docstring in magic method
+    "D107", # Ignore D107 Missing docstring in __init__
+    "D203", # Ignore D203 1 blank line required before class docstring
+    "D204", # Ignore D204 1 blank line required after class docstring
+    "D213", # Ignore D213 Multi-line docstring summary should start at the second line
+    "D215", # Ignore D215 Section underline is over-indented
+    "D400", # Ignore D400 First line should end with a period
+    "D403", # Ignore D403 First word of the first line should be capitalized
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply pyproject.toml]: modified in 484fedacf91212f8c0c9003245ce79b0e1d21841

File: pyproject.toml [hunk 10]
```diff
@@ -85,34 +77,50 @@ log_cli_level = "INFO"
...
+    "D404", # Ignore D404 First word of the docstring should not be "This"
+    "D406", # Ignore D406 Section name should end with a newline
+    "D407", # Ignore D407 Missing dashed underline after section
+    "D408", # Ignore D408 Section underline should be in the line following section's name
+    "D409", # Ignore D409 Section underline should match the length of its name
+    "D413", # Ignore D413 Missing blank line after last section
+    "D415", # Ignore D415 First line should end with a period, question mark, or exclamation point
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply pyproject.toml]: modified in 484fedacf91212f8c0c9003245ce79b0e1d21841

File: pyproject.toml [hunk 11]
```diff
@@ -85,34 +77,50 @@ log_cli_level = "INFO"
...
-# Ignore N818 Exception should be named with an Error suffix
-# Ignore D102 Missing docstring in public method (pydocstyle doesn't look for docstrings in super class
-# Ignore W505 So that strings in comments aren't split across lines
-# https://github.com/PyCQA/pydocstyle/issues/309) TODO: add pylint check? https://github.com/PyCQA/pydocstyle/issues/309#issuecomment-1284142716
-ignore = ["W503", "E501", "D107", "D105", "D415", "D403", "N818", "D102", "W505"]
+[tool.ruff.lint]
+explicit-preview-rules = true
+select = ["A", "E", "W", "F", "C", "N", "D", "I", "CPY001"]
+ignore = [
+    "D102", # Ignore D102 Missing docstring in public method
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply pyproject.toml]: added comments in 484fedacf91212f8c0c9003245ce79b0e1d21841

File: pyproject.toml [hunk 12]
```diff
@@ -85,34 +77,50 @@ log_cli_level = "INFO"
...
+    "D105", # Ignore D105 Missing docstring in magic method
+    "D107", # Ignore D107 Missing docstring in __init__
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply pyproject.toml]: modified in 484fedacf91212f8c0c9003245ce79b0e1d21841

File: pyproject.toml [hunk 13]
```diff
@@ -85,34 +77,50 @@ log_cli_level = "INFO"
...
+    "D400", # Ignore D400 First line should end with a period
+    "D403", # Ignore D403 First word of the first line should be capitalized
+    "D404", # Ignore D404 First word of the docstring should not be "This"
+    "D406", # Ignore D406 Section name should end with a newline
+    "D407", # Ignore D407 Missing dashed underline after section
+    "D408", # Ignore D408 Section underline should be in the line following section's name
+    "D409", # Ignore D409 Section underline should match the length of its name
+    "D413", # Ignore D413 Missing blank line after last section
+    "D415", # Ignore D415 First line should end with a period, question mark, or exclamation point
+    "E501", # Ignore E501 Line too long
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply pyproject.toml]: modified in 484fedacf91212f8c0c9003245ce79b0e1d21841

File: pyproject.toml [hunk 14]
```diff
@@ -85,34 +77,50 @@ log_cli_level = "INFO"
...
+    "N801", # Ignore N801 Class name should use CapWords convention
+    "N818", # Ignore N818 Exception name should be named with an Error suffix
+    "W505", # Ignore W505 Doc line too long
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply pyproject.toml]: modified in 484fedacf91212f8c0c9003245ce79b0e1d21841

File: src/server_exceptions.py [hunk 15]
```diff
@@ -12,7 +12,7 @@ class Error(status_exception.StatusException):
     """MySQL Server unreachable or unhealthy"""
 
 
-class ConnectionError(Error):
+class ConnectionError_(Error):  # noqa: N801
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply src/server_exceptions.py]: addressed in [07664fb](https://github.com/canonical/mysql-router-operator/pull/162/commits/07664fb5f6d833c963c4a14c3e222e5d82876490)

**Comment by taurus-forever** [comment]
@shayancanonical can/should this be merged (after conflicts solving :-( ).

**Pull request merged**